### PR TITLE
fix: redundant CLI downloads [HEAD-450][HEAD-504][HEAD-505]

### PR DESCRIPTION
**Pull request by asaf92**

### CODE REVIEW

File: snykTask/src/install/index.ts [hunk 1]
```diff
@@ -54,17 +54,42 @@ export async function downloadExecutable(
...
+        console.error(`Download of ${executable.filename} failed: ${err.message}`);
+        reject(err);
+      });
       response.pipe(fileWriter);
     });
   });
-}
+
+  // Try to download the file, retry once after 5 seconds if the first attempt fails
+  try {
```
Comment (PeterSchafer) [inline snykTask/src/install/index.ts:82]: Suggestion: How about doing this in a loop?

File: snykTask/src/install/index.ts [hunk 2]
```diff
@@ -54,17 +54,42 @@ export async function downloadExecutable(
...
+    mode: 0o766,
+  });
+
+  // Wrapping the download in a function for easy retrying
+  const doDownload = () => new Promise<void>((resolve, reject) => {
     https.get(executable.downloadUrl, (response) => {
       response.on('end', () => resolve());
-      response.on('error', (err) => reject(err));
+      response.on('error', (err) => {
+        console.error(`Download of ${executable.filename} failed: ${err.message}`);
```
Comment (PeterSchafer) [inline snykTask/src/install/index.ts]: Nitpick: This message is printed twice, here and in the catch clauses.

File: snykTask/src/install/index.ts [hunk 3]
```diff
@@ -54,17 +54,51 @@ export async function downloadExecutable(
...
-  return new Promise<void>((resolve, reject) => {
-    https.get(executable.downloadUrl, (response) => {
-      response.on('end', () => resolve());
-      response.on('error', (err) => reject(err));
-      response.pipe(fileWriter);
-    });
+  const filePath = path.join(targetDirectory, executable.filename);
+
+  // Check if the file already exists
+  if (fs.existsSync(filePath)) {
```
Comment (PeterSchafer) [inline snykTask/src/install/index.ts:60]: Question: Don't we need to compare the latest version and the current version?

Comment (asaf92) [reply snykTask/src/install/index.ts:60]: The files don't survive between pipeline runs so we don't need to worry about it

File: snykTask/src/install/index.ts [hunk 4]
```diff
@@ -53,18 +53,59 @@ export function getSnykDownloadInfo(platform: Platform): SnykDownloads {
...
+    mode: 0o766,
   });
+
+  // Wrapping the download in a function for easy retrying
+  const doDownload = () =>
+    new Promise<void>((resolve, reject) => {
+      https.get(executable.downloadUrl, (response) => {
+        response.on('end', () => resolve());
+        response.on('error', (err) => {
+          console.error(
```
Comment (PeterSchafer) [inline snykTask/src/install/index.ts:78]: nitpick: this log message happens twice, maybe remove one?!

File: snykTask/src/install/index.ts [hunk 5]
```diff
@@ -53,18 +53,59 @@ export function getSnykDownloadInfo(platform: Platform): SnykDownloads {
...
+          reject(err);
+        });
+        response.pipe(fileWriter);
+      });
+    });
+
+  // Try to download the file, retry up to `maxRetries` times if the attempt fails
+  for (let attempt = 0; attempt < maxRetries; attempt++) {
+    try {
+      await doDownload();
```
Comment (PeterSchafer) [inline snykTask/src/install/index.ts:90]: suggestion: let's add a log before doDownload() so that we know when it starts.

Comment (PeterSchafer) [reply snykTask/src/install/index.ts:90]: Maybe we can also include the attempt in the message like
> [0] Start downloading snyk-win.exe
> [0] Download successful for snyk-win.exe

**Comment by asaf92** [comment]
@jessehouwing 
> Have you folks looked into the official tool installer and the tool caching directory for the Azure Pipelines Agent?

I'm not familiar with neither, but it sounds like there's a way to save us some cloud costs so I'm interested!
I assume pipeline caching can be a solution for the CLI downloads, but what's the "official tool installer"?

**Comment by jessehouwing** [comment]
Tool lib is here:

https://github.com/microsoft/azure-pipelines-tool-lib

It provides a set of standard APIs and a naming convention to store tools on the agent in a way that allows multiple jobs to use them, but also shields multiple agents installed on the same machine from eachother.

Example here:

https://github.com/microsoft/azure-devops-extension-tasks/tree/main/BuildTasks/TfxInstaller/v4

In this case the tool installer was broken out into a separate task, but you can merge the 2 together.

The tool installer can optionally be extended with the caching features native to Azure DevOps to store a copy of the tool in the Organization's cache to speed up downloads and reduce bandwidth.

The agent has a build-in caching plugin, but I'm unsure how to leverage this from a custom task outside of the built-in `Cache@2` tasks.

**Comment by bastiandoetsch** [comment]
Do we actually use GZIP compression on the downloads here? Would be good, too :).

**Review by PeterSchafer — APPROVED** [review]
Let's see if these changes fix the immediate issues we are seeing and add an improvement task for using the proposed tooling support.

Thanks a lot @jessehouwing for bringing this to our attention!

**Comment by jessehouwing** [comment]
@PeterSchafer ✅

I also added 2 PRs some time back to improve the representation of the extension on the marketplace.

* #154 
* #155 

I think they may have gotten lost in the team transfer.

**Pull request merged**